### PR TITLE
Validate SSH authorized keys and write the file atomically

### PR DESCRIPTION
**Pull request by agners**
AddSSHAuthKey appended its argument verbatim to /root/.ssh/authorized_keys with no validation. The Supervisor is growing an API on top of this method, which validates user input before calling it, but the method itself should uphold the file format: it is the process actually writing the file as root, and it is reachable by any D-Bus caller the bus policy admits.

Reject input that is not a single well-formed authorized_keys entry, using the OpenSSH parser from golang.org/x/crypto/ssh. Control characters are rejected outright so one call can never write more than one line (newline injection). This is a format check, not policy: options-prefixed entries and any key type the parser accepts still pass.

While at it, harden the write path: create /root/.ssh with mode 0700 when missing (previously the call just failed), replace the O_APPEND write on the live file with a read-append-rename through atomic.WriteFile, and tighten the file mode to 0600 (previously 0644, also for pre-existing files).

Also fix an inverted error check in ClearSSHAuthKeys: it returned an error only when the file was already absent (which should be success) and swallowed genuine deletion failures.

## Summary by CodeRabbit

* **Security Improvements**
  * SSH authorized keys are now strictly validated (whitespace trimming, comment handling, format correctness) and rejected if empty, malformed, contain control characters, or exceed limits.
  * Authorized keys are added using safe, atomic updates with `0600` permissions.
  * Clearing SSH keys now succeeds when the authorized-keys file is already missing.
* **Bug Fixes**
  * Improved newline handling and prevented lost updates during concurrent key additions.
* **Tests**
  * Expanded unit tests for validation edge cases, append behavior, permission enforcement, concurrency, and clear/remove scenarios.
* **Maintenance**
  * Updated indirect Go dependencies.

### CODE REVIEW

File: system/system.go [hunk 1]
```diff
@@ -54,18 +59,68 @@ func (d system) ScheduleWipeDevice() (bool, *dbus.Error) {
...
-	file, err := os.OpenFile(sshAuthKeyFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	_, _, _, rest, err := ssh.ParseAuthorizedKey([]byte(key))
 	if err != nil {
-		logging.Error.Printf("Failed to open SSH authentication file %s: %s", sshAuthKeyFileName, err)
-		return dbus.MakeFailedError(err)
+		return "", fmt.Errorf("invalid SSH authorized key: %w", err)
+	}
+	if len(rest) > 0 {
+		return "", errors.New("unexpected data after SSH authorized key")
+	}
```
[the remark this answers is not in the document]
Comment (agners) [reply system/system.go:99]: The Go convention (staticcheck ST1005) explicitly exempts strings beginning with initialisms and proper nouns; "SSH authorized key is empty" is idiomatic. Rewording to avoid a leading "SSH" would make the messages worse, not more consistent.

File: system/system.go [hunk 2]
```diff
@@ -54,18 +59,68 @@ func (d system) ScheduleWipeDevice() (bool, *dbus.Error) {
...
-	defer file.Close()
+	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
+		return fmt.Errorf("failed to create SSH configuration directory: %w", err)
+	}
+
+	content, err := os.ReadFile(path)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to read SSH authorized keys file: %w", err)
+	}
+	if len(content) > 0 && !bytes.HasSuffix(content, []byte("\n")) {
```
[the remark this answers is not in the document]
Comment (agners) [reply system/system.go:121]: Yeah this seems overengineered.

File: system/system.go [hunk 3]
```diff
@@ -54,18 +59,69 @@ func (d system) ScheduleWipeDevice() (bool, *dbus.Error) {
...
+// accepts (including options) passes. Control characters other than tab,
+// which sshd treats as a field separator, are rejected so a single call can
+// never write more than one line.
+func validateSSHAuthKey(newKey string) (string, error) {
+	key := strings.TrimSpace(newKey)
+	if key == "" {
+		return "", errors.New("SSH authorized key is empty")
+	}
+	for _, r := range key {
+		if (r < 0x20 && r != '\t') || r == 0x7f {
```
Comment (agners) [inline system/system.go]: **Tab-separated entries validate but can never authenticate under dropbear.**

The `'\t'` exemption (and the “tab separators” test case) matches OpenSSH sshd and x/crypto, which both split fields on space *or* tab. Dropbear does not: it requires a single literal space between the key type and the base64 blob ([`svr-authpubkey.c:393`](https://github.com/mkj/dropbear/blob/master/src/svr-authpubkey.c#L393) — `if (buf_getbyte(line) != ' ')`), and its base64 scan terminates only on space (`:401`), so a tab before the comment corrupts the base64 comparison too. Tabs are only tolerated before/inside the options field.

Since HAOS's consumer of this file is dropbear, a tab-separated entry passes validation, gets written, and then silently never works. Suggest dropping the tab exemption (`r < 0x20 || r == 0x7f`) and removing the tab test case — real keys use spaces, and this guarantees every accepted entry is actually usable by dropbear.

File: system/system.go [hunk 4]
```diff
@@ -54,18 +59,69 @@ func (d system) ScheduleWipeDevice() (bool, *dbus.Error) {
...
+			return "", errors.New("SSH authorized key contains control characters")
+		}
+	}
 
-	file, err := os.OpenFile(sshAuthKeyFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	_, _, _, rest, err := ssh.ParseAuthorizedKey([]byte(key))
```
Comment (agners) [inline system/system.go:93]: **No length cap — one oversized entry can disable every key after it for dropbear.**

Dropbear skips lines longer than 3000 bytes (`MAX_AUTHKEYS_LINE`), and [`buf_getline`](https://github.com/mkj/dropbear/blob/master/src/dbutil.c#L551) returns failure for lines over `DROPBEAR_MAX_LINE_LENGTH` (10000), which `checkpubkey` treats as EOF — so a single huge entry makes all subsequent lines in the file unreachable (it also caps at 300 lines total).

The parser here accepts arbitrarily long comments, so a valid-but-huge entry passes. Suggest rejecting entries longer than ~3000 bytes to match dropbear's per-line limit (a 16384-bit RSA line is ~2786 bytes, so real keys fit).

**Review by agners — COMMENTED** [review]
Reviewed with a focus on dropbear compatibility, since on HAOS the consumer of `/root/.ssh/authorized_keys` is dropbear (`dropbear.service.d/haos.conf`, port 22222), not OpenSSH sshd. Compared the validator against dropbear's parser (`src/svr-authpubkey.c`, `src/dbutil.c`). Everything the validator rejects is also invalid for dropbear, so no false rejections — but two accept-but-unusable cases below.

File: system/system_test.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+	var wg sync.WaitGroup
+	errs := make([]error, workers)
+	for i := range workers {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			errs[i] = addSSHAuthKey(path, fmt.Sprintf("%s worker-%d", testKeyEd25519, i))
+		}()
+	}
```
[the remark this answers is not in the document]
Comment (agners) [reply system/system_test.go:171]: Well, it does compile, and is valid since Go 1.22.

**Review by mdegat01 — APPROVED** [review]
Nice catch, LGTM 👍
